File: bench/cpp/ops/BenchLabel.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchLabelConfig.hpp"

#include <cvcuda/OpLabel.hpp>
#include <cvcuda/OpReformat.hpp>

#include <nvbench/nvbench.cuh>

#include <string>
#include <string_view>
#include <utility>

inline bool HasRunChoice(std::string_view runChoice, std::string_view choice)
{
    return runChoice.find(choice) != std::string::npos;
}

struct LabelLayout
{
    std::string runChoice;
    bool        isPlanar        = false;
    bool        isFakePlanar    = false;
    bool        usePlanarLayout = false;
    bool        valid           = true;
};

template<typename ST, typename DT>
struct LabelTensors
{
    nvcv::Tensor src;
    nvcv::Tensor dst;
};

inline LabelLayout ParseLabelLayout(nvbench::state &state, std::string runChoice)
{
    if (runChoice == "DEFAULT")
    {
        runChoice = "";
    }

    const std::string layout = benchutils::GetStringParam(state, "layout", "NHWC");
    if (layout != "NHWC" && layout != "NCHW" && layout != "NCHW_FAKE")
    {
        state.skip("Label benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return {"", false, false, false, false};
    }

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";

    return {std::move(runChoice), isPlanar, isFakePlanar, isPlanar || isFakePlanar, true};
}

template<typename ST, typename DT>
void SetupOptionalLabelInputs(const std::string &runChoice, long3 srcShape, long3 staShape, bool usePlanarLayout,
                              nvcv::Tensor &bgT, nvcv::Tensor &minT, nvcv::Tensor &maxT, nvcv::Tensor &mszT,
                              nvcv::Tensor &countT, nvcv::Tensor &statsT, nvcv::Tensor &maskT)
{
    if (HasRunChoice(runChoice, "BG"))
    {
        bgT = nvcv::Tensor({{srcShape.x}, "N"}, benchutils::GetDataType<ST>());

        benchutils::FillTensor<ST>(bgT, benchutils::LcgValues<ST>());
    }
    if (HasRunChoice(runChoice, "MIN"))
    {
        minT = nvcv::Tensor({{srcShape.x}, "N"}, benchutils::GetDataType<ST>());

        const ST minThreshold = std::is_signed_v<ST> ? static_cast<ST>(-64) : static_cast<ST>(64);
        benchutils::FillTensor<ST>(minT, benchutils::LcgValues<ST>(minThreshold, minThreshold));
    }
    if (HasRunChoice(runChoice, "MAX"))
    {
        maxT = nvcv::Tensor({{srcShape.x}, "N"}, benchutils::GetDataType<ST>());

        const ST maxThreshold = std::is_signed_v<ST> ? static_cast<ST>(64) : static_cast<ST>(192);
        benchutils::FillTensor<ST>(maxT, benchutils::LcgValues<ST>(maxThreshold, maxThreshold));
    }
    if (HasRunChoice(runChoice, "ISLAND"))
    {
        mszT = nvcv::Tensor({{srcShape.x}, "N"}, benchutils::GetDataType<DT>());

        benchutils::FillTensor<DT>(mszT, benchutils::LcgValues<DT>(16, 16));
    }
    if (HasRunChoice(runChoice, "COUNT"))
    {
        countT = nvcv::Tensor({{srcShape.x}, "N"}, benchutils::GetDataType<DT>());
    }
    if (HasRunChoice(runChoice, "STAT"))
    {
        statsT = nvcv::Tensor(
            {
                {staShape.x, staShape.y, staShape.z},
                "NMA"
        },
            benchutils::GetDataType<DT>());
    }
    if (HasRunChoice(runChoice, "MASK"))
    {
        maskT = usePlanarLayout ? nvcv::Tensor(
                    {
                        {srcShape.x, 1, srcShape.y, srcShape.z},
                        "NCHW"
        },
                    nvcv::TYPE_U8)
                                : nvcv::Tensor({{srcShape.x, srcShape.y, srcShape.z, 1}, "NHWC"}, nvcv::TYPE_U8);

        benchutils::FillTensor<uint8_t>(maskT, benchutils::CheckerboardValues<uint8_t>());
    }
}

template<typename ST, typename DT>
LabelTensors<ST, DT> BuildLabelTensors(long3 srcShape, long3 dstShape, bool usePlanarLayout)
{
    if (usePlanarLayout)
    {
        return {
            nvcv::Tensor({{srcShape.x, 1, srcShape.y, srcShape.z}, "NCHW"},
            benchutils::GetDataType<ST>()),
            nvcv::Tensor({{dstShape.x, 1, dstShape.y, dstShape.z}, "NCHW"},
            benchutils::GetDataType<DT>())
        };
    }

    return {
        nvcv::Tensor({{srcShape.x, srcShape.y, srcShape.z, 1}, "NHWC"},
        benchutils::GetDataType<ST>()),
        nvcv::Tensor({{dstShape.x, dstShape.y, dstShape.z, 1}, "NHWC"},
        benchutils::GetDataType<DT>())
    };
}

template<typename ST, typename DT>
void RunLabelFakePlanar(nvbench::state &state, cvcuda::Label &op, cvcuda::Reformat &reformatOp, nvcv::Tensor &src,
                        nvcv::Tensor &dst, long3 srcShape, long3 dstShape, nvcv::Tensor &bgT, nvcv::Tensor &minT,
                        nvcv::Tensor &maxT, nvcv::Tensor &mszT, nvcv::Tensor &countT, nvcv::Tensor &statsT,
                        nvcv::Tensor &maskT, NVCVConnectivityType conn, NVCVLabelType alab, NVCVLabelMaskType mType)
{
    nvcv::Tensor interSrc(
        {
            {srcShape.x, srcShape.y, srcShape.z, 1},
            "NHWC"
    },
        benchutils::GetDataType<ST>());
    nvcv::Tensor interDst(
        {
            {dstShape.x, dstShape.y, dstShape.z, 1},
            "NHWC"
    },
        benchutils::GetDataType<DT>());
    nvcv::Tensor interMaskT;
    if (maskT)
    {
        interMaskT = nvcv::Tensor(
            {
                {srcShape.x, srcShape.y, srcShape.z, 1},
                "NHWC"
        },
            nvcv::TYPE_U8);
    }

    benchutils::warmup_and_exec(state, BENCH_LABEL_WARMUP_ITERATIONS,
                                [&op, &reformatOp, &src, &dst, &interSrc, &interDst, &bgT, &minT, &maxT, &mszT, &countT,
                                 &statsT, &maskT, &interMaskT, &conn, &alab, &mType](cudaStream_t s)
                                {
                                    reformatOp(s, src, interSrc);
                                    if (maskT)
                                    {
                                        reformatOp(s, maskT, interMaskT);
                                    }
                                    op(s, interSrc, interDst, bgT, minT, maxT, mszT, countT, statsT, interMaskT, conn,
                                       alab, mType);
                                    reformatOp(s, interDst, dst);
                                });
}

template<typename ST>
inline void label(nvbench::state &state, nvbench::type_list<ST>)
try
{
    // Use int (S32) for label output dtype to match Python benchmark
    using DT = int;

    long3 srcShape = benchutils::GetShape<3>(state.get_string("shape"));
    long3 dstShape = srcShape;

    LabelLayout layout = ParseLabelLayout(state, state.get_string("runChoice"));
    if (!layout.valid)
    {
        return;
    }

    // Use [BG][MIN][MAX][ISLAND][COUNT][STAT][MASK] in runChoice to run Label with:
    // background; minThreshold; maxThreshold; island removal; count; statistics; mask

    long3 staShape{srcShape.x, 10000, 7}; // using fixed 10K max. cap. and 2D problem

    NVCVConnectivityType conn  = NVCV_CONNECTIVITY_4_2D;
    NVCVLabelType        alab  = NVCV_LABEL_FAST;
    NVCVLabelMaskType    mType = NVCV_REMOVE_ISLANDS_OUTSIDE_MASK_ONLY;

    nvcv::Tensor bgT;
    nvcv::Tensor minT;
    nvcv::Tensor maxT;
    nvcv::Tensor countT;
    nvcv::Tensor statsT;
    nvcv::Tensor mszT;
    nvcv::Tensor maskT;

    cvcuda::Label op;

    const long imageBytes = srcShape.x * srcShape.y * srcShape.z * sizeof(ST);
    const long outBytes   = dstShape.x * dstShape.y * dstShape.z * sizeof(DT);
    if (layout.isFakePlanar)
    {
        state.add_global_memory_reads(2 * imageBytes + outBytes);
        state.add_global_memory_writes(imageBytes + 2 * outBytes);
    }
    else
    {
        state.add_global_memory_reads(imageBytes);
        state.add_global_memory_writes(outBytes);
    }

    SetupOptionalLabelInputs<ST, DT>(layout.runChoice, srcShape, staShape, layout.usePlanarLayout, bgT, minT, maxT,
                                     mszT, countT, statsT, maskT);

    LabelTensors<ST, DT> tensors = BuildLabelTensors<ST, DT>(srcShape, dstShape, layout.usePlanarLayout);

    benchutils::FillTensor<ST>(tensors.src, benchutils::LcgValues<ST>());

    if (layout.isFakePlanar)
    {
        cvcuda::Reformat reformatOp;
        RunLabelFakePlanar<ST, DT>(state, op, reformatOp, tensors.src, tensors.dst, srcShape, dstShape, bgT, minT, maxT,
                                   mszT, countT, statsT, maskT, conn, alab, mType);
    }
    else
    {
        benchutils::warmup_and_exec(
            state, BENCH_LABEL_WARMUP_ITERATIONS,
            [&op, &src = tensors.src, &dst = tensors.dst, &bgT, &minT, &maxT, &mszT, &countT, &statsT, &maskT, &conn,
             &alab, &mType](cudaStream_t s)
            { op(s, src, dst, bgT, minT, maxT, mszT, countT, statsT, maskT, conn, alab, mType); });
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(label, NVBENCH_TYPE_AXES(BENCH_LABEL_TYPES))
BENCH_LABEL_AXES;

File: bench/python/ops/bench_label.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA Label operator benchmark - Python equivalent of BenchLabel.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    parse_shape,
    get_dtype,
    get_dtype_size,
    create_tensor,
    create_stream_cache,
    run_benchmark,
)


def label(state):
    """Label operator benchmark matching C++ BenchLabel.cpp"""

    N, H, W = parse_shape(state.get_string("shape"))
    dtype = get_dtype(state.get_string("InOutDataType"))
    runChoice = state.get_string("runChoice")
    device_id = state.get_device()
    try:
        layout = state.get_string("layout")
    except (KeyError, RuntimeError):
        layout = "NHWC"

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip("Label benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts")
        return None

    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"

    if runChoice == "DEFAULT":
        runChoice = ""

    dtype_size = get_dtype_size(dtype)
    image_bytes = N * H * W * dtype_size
    out_bytes = N * H * W * 4
    if is_fake_planar:
        state.add_global_memory_reads(2 * image_bytes + out_bytes)
        state.add_global_memory_writes(image_bytes + 2 * out_bytes)
    else:
        state.add_global_memory_reads(image_bytes)
        state.add_global_memory_writes(out_bytes)

    tensor_shape = (N, 1, H, W) if is_planar or is_fake_planar else (N, H, W, 1)
    tensor_layout = "NCHW" if is_planar or is_fake_planar else "NHWC"

    src = create_tensor(
        tensor_shape, dtype, device_id, layout=tensor_layout, fill_mode="lcg"
    )

    dst = create_tensor(
        tensor_shape, cvcuda.Type.S32, device_id, layout=tensor_layout, fill_mode=0
    )

    bgLabel = None
    minThresh = None
    maxThresh = None
    minSize = None
    mask = None

    if "BG" in runChoice:
        bgLabel = create_tensor((N,), dtype, device_id, layout="N", fill_mode="lcg")
    if "MIN" in runChoice:
        minThresh = create_tensor((N,), dtype, device_id, layout="N", fill_mode=64)
    if "MAX" in runChoice:
        maxThresh = create_tensor((N,), dtype, device_id, layout="N", fill_mode=192)
    if "ISLAND" in runChoice:
        minSize = create_tensor(
            (N,), cvcuda.Type.S32, device_id, layout="N", fill_mode=16
        )
    if "MASK" in runChoice:
        mask = create_tensor(
            tensor_shape,
            cvcuda.Type.U8,
            device_id,
            layout=tensor_layout,
            fill_mode="checkerboard",
        )

    countTensor = None
    statsTensor = None

    if "COUNT" in runChoice:
        countTensor = create_tensor(
            (N,), cvcuda.Type.S32, device_id, layout="N", fill_mode=0
        )
    if "STAT" in runChoice:
        statsTensor = create_tensor(
            (N, 10000, 7), cvcuda.Type.S32, device_id, layout="NMA", fill_mode=0
        )

    conn = cvcuda.ConnectivityType.CONNECTIVITY_4_2D
    alab = cvcuda.LABEL.FAST
    mType = cvcuda.LabelMaskType.REMOVE_ISLANDS_OUTSIDE_MASK_ONLY

    get_stream = create_stream_cache()

    if is_fake_planar:
        inter_src = create_tensor(
            (N, H, W, 1), dtype, device_id, layout="NHWC", fill_mode=0
        )
        inter_dst = create_tensor(
            (N, H, W, 1), cvcuda.Type.S32, device_id, layout="NHWC", fill_mode=0
        )
        inter_mask = None
        if mask is not None:
            inter_mask = create_tensor(
                (N, H, W, 1), cvcuda.Type.U8, device_id, layout="NHWC", fill_mode=0
            )

        def run_fake(launch):
            stream = get_stream(launch)
            cvcuda.reformat_into(inter_src, src, stream=stream)
            if mask is not None:
                cvcuda.reformat_into(inter_mask, mask, stream=stream)
            cvcuda.label_into(
                inter_dst,
                countTensor,
                statsTensor,
                inter_src,
                connectivity=conn,
                assign_labels=alab,
                mask_type=mType,
                bg_label=bgLabel,
                min_thresh=minThresh,
                max_thresh=maxThresh,
                min_size=minSize,
                mask=inter_mask,
                stream=stream,
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    def run(launch):
        cvcuda.label_into(
            dst,
            countTensor,
            statsTensor,
            src,
            connectivity=conn,
            assign_labels=alab,
            mask_type=mType,
            bg_label=bgLabel,
            min_thresh=minThresh,
            max_thresh=maxThresh,
            min_size=minSize,
            mask=mask,
            stream=get_stream(launch),
        )

    return run


if __name__ == "__main__":
    run_benchmark("label", label)
